SYCL/ESIMD/api/functional/functions/functions_select_rvalue.hpp

```diff
@@ -0,0 +1,204 @@
+  }
+};
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename TestCaseT, typename NumSelectedElemsT, typename StrideT,
+          typename UseOffsetT, typename DataT, typename DimT>
+class run_test {
+  static constexpr int NumElems = DimT::value;
+  static constexpr int NumSelectedElems = NumSelectedElemsT::value;
+  static constexpr int Stride = StrideT::value;
+  static constexpr bool UseOffset = UseOffsetT::value;
+
+public:
+  bool operator()(sycl::queue &queue, size_t offset,
+                  const std::string &data_type) {
+
+    bool passed = true;
+    size_t alignment_value = alignof(DataT);
+
+    constexpr size_t value_for_increase_ref_data_for_fill = 50;
+    static_assert(std::numeric_limits<char>::max() >
+                      value_for_increase_ref_data_for_fill + NumElems,
+                  "Value that used for increase ref data for fill plus simd "
+                  "size should be less than char max value.");
+
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> result(NumElems, allocator);
+    shared_vector<DataT> initial_ref_data(NumElems, allocator);
+    shared_vector<DataT> ref_data_for_fill(NumElems, allocator);
+
+    shared_element<bool> is_correct_type(queue, true);
+
+    for (size_t i = 0; i < NumElems; ++i) {
+      initial_ref_data[i] = i + 1;
+    }
+    // We should have different values in the first reference data and in the
+    // second reference data.
+    for (size_t i = 0; i < NumElems; ++i) {
+      ref_data_for_fill[i] =
+          initial_ref_data[i] + value_for_increase_ref_data_for_fill;
+    }
+
+    queue.submit([&](sycl::handler &cgh) {
+      DataT *init_ref_ptr = initial_ref_data.data();
+      DataT *ref_data_for_fill_ptr = ref_data_for_fill.data();
+      DataT *const out_ptr = result.data();
+      auto is_correct_type_ptr = is_correct_type.data();
+
+      cgh.single_task<
+          Kernel<DataT, NumElems, TestCaseT, NumSelectedElemsT, StrideT>>(
+          [=]() SYCL_ESIMD_KERNEL {
+            *is_correct_type_ptr =
+                TestCaseT::template call_operator<DataT, NumElems,
+                                                  NumSelectedElems, Stride>(
+                    init_ref_ptr, ref_data_for_fill_ptr, out_ptr, offset);
+          });
+    });
+    queue.wait_and_throw();
+
+    // Verify that selected values has been changed to values from referece data
+    // for fill.
+    for (size_t i = offset; i < Stride * NumSelectedElems + offset;
+         i += Stride) {
+      DataT expected = ref_data_for_fill[(i - offset) / Stride];
+      DataT retrieved = result[i];
+      if (expected != retrieved) {
+        passed = fail_test(i, expected, retrieved, data_type);
+      }
+    }
+    // Verify that values, that do not was selected has initial values.
+    for (size_t i = 0; i < NumElems; ++i) {
+      if (i % Stride != 0) {
```

Comment:
You've actually had presented a great approach with `O(N)` complexity for a different PR. Why not to use it here as well?

Reply:
The new algorithm has been implemented [here](https://github.com/intel/llvm-test-suite/pull/875/commits/9519eccf663e28b1760cf5bf869ba75ea5fa275d#diff-ee52ba82babf88cc257b75cc71c61eca3df7aa19a2b6ce7b5ed19349642a56f6R134-R155)